src/broker/overlay.c

```diff
@@ -426,6 +426,29 @@ static struct child *child_lookup (struct overlay *ov, const char *id)
     return NULL;
 }
 
+bool overlay_request_is_local (struct overlay *ov, const flux_msg_t *msg)
+{
+    int type;
+    const char *route;
+
+    if (ov == NULL
+        || flux_msg_get_type (msg, &type) < 0
+        || type != FLUX_MSGTYPE_REQUEST
+        || !(route = flux_msg_route_last (msg))) {
+        errno = EINVAL;
+        return false;
+    }
+
+    /*  If the last pushed route is a peer or parent, then this message
+     *   was definitely not local.
+     */
+    if (strcmp (route, ov->parent.uuid) == 0
+        || child_lookup_online (ov, route)
+        || child_lookup (ov, route))
```

Comment:
`child_lookup_online()` could be dropped if `child_lookup()` is called.

Reply:
I was thinking that would be faster most of the time, but that was probably dumb. Sorry.

Reply:
Oh sorry, I didn't think of that.  Maybe just add a comment to that effect?

Reply:
Ok, I added that comment.

It also occurred to me another way to do this might be to flag messages that are received from the overlay (e.g. `flux_msg_aux_set (msg, "overlay::is_remote", int2ptr (1), NULL)`. This would avoid a guaranteed loop through all children on every local message, though I'm not sure that's a real issue in this limited case.

Reply:
Since the implementation is simple: here's what would be the diff between what we have here and using an aux "flag" on messages. I'm not sure, though, if this is premature optimization, or if it is even a fully sufficient solution. What do you think? (dumb tests pass anyway)

```diff
diff --git a/src/broker/overlay.c b/src/broker/overlay.c
index e5d093f70..99f67fc8e 100644
--- a/src/broker/overlay.c
+++ b/src/broker/overlay.c
@@ -31,6 +31,7 @@
 #include "src/common/libutil/errno_safe.h"
 #include "src/common/libutil/monotime.h"
 #include "src/common/librouter/rpc_track.h"
+#include "src/common/libccan/ccan/ptrint/ptrint.h"
 
 #include "overlay.h"
 #include "attr.h"
@@ -438,15 +439,7 @@ bool overlay_request_is_local (struct overlay *ov, const flux_msg_t *msg)
         errno = EINVAL;
         return false;
     }
-
-    /*  If the last pushed route is a peer or parent, then this message
-     *   was definitely not local. Note: we call child_lookup_online()
-     *   before child_lookup() becuase the former is potentially much
-     *   faster.
-     */
-    if (strcmp (route, ov->parent.uuid) == 0
-        || child_lookup_online (ov, route)
-        || child_lookup (ov, route))
+    if (flux_msg_aux_get (msg, "overlay::remote"))
         return false;
     return true;
 }
@@ -881,6 +874,8 @@ static void child_cb (flux_reactor_t *r, flux_watcher_t *w,
 
     if (!(msg = zmqutil_msg_recv (ov->bind_zsock)))
         return;
+    if (flux_msg_aux_set (msg, "overlay::remote", int2ptr (1), NULL) < 0)
+        goto done;
     if (flux_msg_get_type (msg, &type) < 0
         || !(uuid = flux_msg_route_last (msg))) {
         logdrop (ov, OVERLAY_DOWNSTREAM, msg, "malformed message");
@@ -975,6 +970,8 @@ static void parent_cb (flux_reactor_t *r, flux_watcher_t *w,
 
     if (!(msg = zmqutil_msg_recv (ov->parent.zsock)))
         return;
+    if (flux_msg_aux_set (msg, "overlay::remote", int2ptr (1), NULL) < 0)
+        goto done;
     if (flux_msg_get_type (msg, &type) < 0) {
         logdrop (ov, OVERLAY_UPSTREAM, msg, "malformed message");
         goto done;
```

Reply:
That seems very close to me!

One problem here though is that the aux container isn't part of the serialized message, so it gets cleared when the message is forwarded (for example to a broker module).   It seems like the right thing to do here would be to define a new message flag in RFC 3 and set it in the flags field of the message "proto frame'?

Reply:
My thought was that currently `overlay_request_is_local()` is only exported to the broker code specifically, so perhaps this limited support is ok for now (and perhaps better than checking the parent and child uuids directly each time?)

When we want to export this functionality to modules or via libflux, then I think the idea of a new flag is great. Does that make sense?

Reply:
Ah yes that does make sense.  OK, this looks good then!

I would suggest adding comments where the message is annotated stating that it's for use by `overlay_request_is_local()`.  Or maybe `overlay_message_is_remote()` would be more appropriate now.

Reply:
Good suggestions!

Reply:
Ok, I made this change and forced a push.

I kept the `is_local()` api for this function. This is because then the function can "fail safe" by returning false if `msg == NULL`.

Also, the function can work on any message now I assumed, not just requests, so it is renamed `overlay_msg_is_local()`. Let me know what you think.